File: node/be/db/steps/cex.steps.js
'use strict';

const fs = require('fs');
const path = require('path');
const { Given, When, Then, Before, After } = require('@cucumber/cucumber');
const { Store, DbUnreachable, throwaway } = require('../store');
const { MiniCex, ApiUnreachable } = require('../../api/venues/minicex');

/**
 * Steps for features/be-cex-db.feature, plus the shared "drive the service"
 * account and action steps that be-cex-api.feature and fe-cex.feature reuse.
 *
 * Isolation is by fresh accounts and deltas, not by reset: each scenario makes
 * its own accounts, drives the API, and asserts what changed -- so scenarios
 * never collide however they are ordered.
 */

const cex = new MiniCex();
const SEED = path.join(__dirname, '..', '..', '..', '..', 'services', 'mini-cex', 'db', 'seed.sql');
const UNREACHABLE = [DbUnreachable, ApiUnreachable];
const coin = (n) => Math.round(Number(n) * 1e8);

Before(function () {
  this.store = null;
  this.cex = cex;
  this.accounts = {};      // alias -> { id, handle, token }
  this.api = null;         // last HTTP response
  this.tmp = null;         // throwaway db
  this.lastBridge = null;  // last bridge op body
  this.quote = null;
});
After(function () {
  if (this.tmp) { try { this.tmp.close(); } catch { /* fine */ } }
  if (this.store) this.store.close();
});

// ---------------------------------------------------------------- helpers

async function act(world, fn) {
  if (world.sourceError) return undefined;
  try { return await fn(); } catch (err) {
    if (UNREACHABLE.some((C) => err instanceof C)) { world.sourceError = err.message; return undefined; }
    throw err;
  }
}
async function check(world, description, fn) {
  if (world.sourceError) { world.unobservable(description, 'the source could not be reached -- ' + world.sourceError); return; }
  let r;
  try { r = await fn(); } catch (err) { if (UNREACHABLE.some((C) => err instanceof C)) { world.unobservable(description, err.message); return; } throw err; }
  world.check(description, r.passed, r.detail);
}
function expectThrow(fn, needle) {
  try { fn(); return { passed: false, detail: 'no error thrown' }; }
  catch (err) { const ok = !needle || new RegExp(needle, 'i').test(err.message); return { passed: ok, detail: ok ? 'rejected: ' + err.message.split('\n')[0] : 'wrong error: ' + err.message }; }
}
async function ensureAccount(world, alias) {
  if (world.accounts[alias]) return world.accounts[alias];
  const a = await cex.createAccount(alias);
  world.accounts[alias] = a;
  return a;
}
const acctOf = (world, alias) => world.accounts[alias];

// ---------------------------------------------------------------- Background

Given('the store is open and the service is reachable', { timeout: 30_000 }, async function () {
  if (this.sourceError) return;
  try { this.store = new Store(); } catch (err) { if (err instanceof DbUnreachable) { this.sourceError = err.message; return; } throw err; }
  const r = await act(this, () => cex.get('/assets'));
  if (this.sourceError) return;
  if (!r || r.status !== 200) { this.sourceError = 'mini-cex /assets returned ' + (r ? r.status : 'nothing'); }
});

// ---------------------------------------------------------------- throwaway schema

Given('a throwaway database with the schema applied', function () {
  this.tmp = throwaway();
  // Seed the parent assets the constraint tests reference, so the constraint
  // under test (a CHECK, a UNIQUE) is what fires -- not an incidental missing
  // asset FOREIGN KEY.
  this.tmp.exec("INSERT INTO assets (symbol, name, kind, decimals, usd_micro) VALUES ('BTC','Bitcoin','crypto',8,1), ('USD','US Dollar','fiat',2,1)");
});
Given('a throwaway database with the schema and seed applied', function () {
  this.tmp = throwaway();
  this.tmp.exec(fs.readFileSync(SEED, 'utf8'));
});

Then(/^the store has tables (.+)$/, async function (csv) {
  await check(this, 'documented tables present', async () => {
    const have = new Set(this.store.tables());
    const want = csv.split(',').map((s) => s.trim());
    const missing = want.filter((t) => !have.has(t));
    return { passed: missing.length === 0, detail: missing.length ? 'missing ' + missing.join(',') : 'all present' };
  });
});
Then('inserting a balance with a negative amount fails a CHECK', function () {
  this.tmp.exec("INSERT INTO accounts (id, handle, created_at) VALUES (1, 'x', 0)");
  const r = expectThrow(() => this.tmp.exec("INSERT INTO balances (account_id, asset, amount) VALUES (1, 'BTC', -1)"), 'CHECK|constraint');
  this.check('negative balance rejected', r.passed, r.detail);
});
Then('inserting a ledger row with zero delta fails a CHECK', function () {
  this.tmp.exec("INSERT INTO accounts (id, handle, created_at) VALUES (1, 'x', 0)");
  const r = expectThrow(() => this.tmp.exec("INSERT INTO ledger (account_id, asset, delta, reason, created_at) VALUES (1, 'BTC', 0, 'deposit', 0)"), 'CHECK|constraint');
  this.check('zero delta rejected', r.passed, r.detail);
});
Then('inserting a ledger row with reason {string} fails a CHECK', function (reason) {
  const r = expectThrow(() => this.tmp.exec(`INSERT INTO ledger (account_id, asset, delta, reason, created_at) VALUES (1, 'BTC', 1, '${reason}', 0)`), 'CHECK|constraint');
  this.check('bad reason rejected', r.passed, r.detail);
});
Then('inserting a balance for a missing account fails a FOREIGN KEY', function () {
  const r = expectThrow(() => this.tmp.exec("INSERT INTO balances (account_id, asset, amount) VALUES (999, 'BTC', 1)"), 'FOREIGN KEY|constraint');
  this.check('missing account FK rejected', r.passed, r.detail);
});
Then('inserting two deposit bridge ops with the same external tx fails on the second', function () {
  this.tmp.exec("INSERT INTO accounts (id, handle, created_at) VALUES (1, 'x', 0)");
  const ins = "INSERT INTO bridge_ops (account_id, asset, amount, direction, src_chain, dst_chain, status, ext_tx, created_at, updated_at) VALUES (1, 'BTC', 1, 'deposit', 'ethereum', 'mini-cex', 'credited', 'tx1', 0, 0)";
  this.tmp.exec(ins);
  const r = expectThrow(() => this.tmp.exec(ins), 'UNIQUE|constraint');
  this.check('duplicate bridge ext_tx rejected', r.passed, r.detail);
});
Then('inserting a pair whose base equals its quote fails a CHECK', function () {
  const r = expectThrow(() => this.tmp.exec("INSERT INTO pairs (base, quote) VALUES ('BTC', 'BTC')"), 'CHECK|constraint');
  this.check('same-leg pair rejected', r.passed, r.detail);
});
Then('applying the seed again changes no row counts', function () {
  const cnt = () => ({ assets: this.tmp.prepare('SELECT COUNT(*) AS n FROM assets').get().n, pairs: this.tmp.prepare('SELECT COUNT(*) AS n FROM pairs').get().n });
  const before = cnt();
  this.tmp.exec(fs.readFileSync(SEED, 'utf8'));
  const after = cnt();
  this.check('seed is idempotent', before.assets === after.assets && before.pairs === after.pairs, `assets ${before.assets}->${after.assets}, pairs ${before.pairs}->${after.pairs}`);
});

// ---------------------------------------------------------------- shared account + action steps

Given('a fresh account {string}', async function (alias) {
  await act(this, () => ensureAccount(this, alias));
});
Given('a fresh account {string} funded with {float} {word}', async function (alias, amt, asset) {
  await act(this, async () => { const a = await ensureAccount(this, alias); this.api = await cex.deposit(a.token, asset, coin(amt)); });
});

async function deposit(world, alias, amt, asset, key) {
  const a = await ensureAccount(world, alias);
  world.api = await cex.deposit(a.token, asset, coin(amt), key);
}
// One definition per phrase: Cucumber shares its registry across Given/When/Then,
// so this single step serves both the "And ... deposits" (Given) setup lines and
// the "When ... deposits" action lines in the features.
When('{string} deposits {float} {word}', async function (alias, amt, asset) { await act(this, () => deposit(this, alias, amt, asset)); });
When('{string} deposits {float} {word} with key {string}', async function (alias, amt, asset, key) { await act(this, () => deposit(this, alias, amt, asset, key)); });

When('{string} withdraws {float} {word}', async function (alias, amt, asset) {
  await act(this, async () => { this.api = await cex.withdraw(acctOf(this, alias).token, asset, coin(amt)); });
});
When('{string} transfers {float} {word} to {string}', async function (alias, amt, asset, to) {
  await act(this, async () => { const toAcct = await ensureAccount(this, to); this.api = await cex.transfer(acctOf(this, alias).token, toAcct.handle, asset, coin(amt)); });
});
When('{string} tries to transfer {float} {word} to {string}', async function (alias, amt, asset, to) {
  await act(this, async () => { const toAcct = await ensureAccount(this, to); this.api = await cex.transfer(acctOf(this, alias).token, toAcct.handle, asset, coin(amt)); });
});
When('{string} swaps {float} {word} to {word}', async function (alias, amt, from, to) {
  await act(this, async () => { this.api = await cex.swap(acctOf(this, alias).token, { from_asset: from, to_asset: to, from_amount: coin(amt), idempotency_key: null }); });
});
When('{string} swaps {float} {word} to {word} demanding at least {float} {word} out', async function (alias, amt, from, to, minv, _q) {
  await act(this, async () => { this.api = await cex.swap(acctOf(this, alias).token, { from_asset: from, to_asset: to, from_amount: coin(amt), min_to_amount: coin(minv) }); });
});
When('{string} bridge-withdraws {float} {word} to chain {string}', async function (alias, amt, asset, chain) {
  await act(this, async () => { this.api = await cex.bridgeWithdraw(acctOf(this, alias).token, { asset, amount: coin(amt), dst_chain: chain }); if (this.api.status === 201) this.lastBridge = this.api.body; });
});
When('a bridge deposit of {float} {word} for {string} from chain {string} with tx {string} is observed', async function (amt, asset, alias, chain, tx) {
  await act(this, async () => { const a = await ensureAccount(this, alias); this.bridgeDepositReq = { handle: a.handle, asset, amount: coin(amt), src_chain: chain, ext_tx: tx }; this.api = await cex.bridgeDeposit(this.bridgeDepositReq); if (this.api.body && this.api.body.id) this.lastBridge = this.api.body; });
});
When('the same bridge deposit is observed again', async function () {
  await act(this, async () => { this.api = await cex.bridgeDeposit(this.bridgeDepositReq); });
});
When('the bridge op is confirmed', async function () {
  await act(this, async () => { this.api = await cex.bridgeConfirm(this.lastBridge.id); this.lastBridge = this.api.body; });
});

// ---------------------------------------------------------------- DB Thens

function bal(world, alias, asset) { const id = world.store.accountId(acctOf(world, alias).handle); return world.store.balance(id, asset); }
function ledgerSum(world, alias, asset) { const id = world.store.accountId(acctOf(world, alias).handle); return world.store.ledgerBalance(id, asset); }

Then('the balance row for {string} in {word} is {float} {word}', async function (alias, asset, amt, _a) {
  await check(this, `balance ${alias}/${asset}`, async () => { const got = bal(this, alias, asset); return { passed: got === coin(amt), detail: `got ${got}, want ${coin(amt)}` }; });
});
Then('a ledger line credits {string} {float} {word} with reason {string}', async function (alias, amt, asset, reason) {
  await check(this, `ledger credit ${reason}`, async () => { const id = this.store.accountId(acctOf(this, alias).handle); const row = this.store.get('SELECT * FROM ledger WHERE account_id = ? AND asset = ? AND delta = ? AND reason = ?', id, asset, coin(amt), reason); return { passed: !!row, detail: row ? 'found' : 'no matching ledger line' }; });
});
Then('a ledger line debits {string} {float} {word} with reason {string}', async function (alias, amt, asset, reason) {
  await check(this, `ledger debit ${reason}`, async () => { const id = this.store.accountId(acctOf(this, alias).handle); const row = this.store.get('SELECT * FROM ledger WHERE account_id = ? AND asset = ? AND delta = ? AND reason = ?', id, asset, -coin(amt), reason); return { passed: !!row, detail: row ? 'found' : 'no matching ledger line' }; });
});
Then('the ledger sum for {string} in {word} equals the balance row', async function (alias, asset) {
  await check(this, `ledger == balance ${asset}`, async () => { const s = ledgerSum(this, alias, asset); const b = bal(this, alias, asset); return { passed: s === b, detail: `ledger ${s}, balance ${b}` }; });
});
Then('the total BTC across {string} and {string} is unchanged', async function (a1, a2) {
  await check(this, 'total conserved', async () => { const t = bal(this, a1, 'BTC') + bal(this, a2, 'BTC'); return { passed: t === coin(1.0), detail: `total ${t}` }; });
});
Then('{string} has a swap row from {word} to {word}', async function (alias, from, to) {
  await check(this, 'swap row present', async () => { const id = this.store.accountId(acctOf(this, alias).handle); const row = this.store.get('SELECT * FROM swaps WHERE account_id = ? AND from_asset = ? AND to_asset = ?', id, from, to); return { passed: !!row, detail: row ? `swap #${row.id}` : 'no swap row' }; });
});
Then('the bridge op for {string} is in status {string}', async function (alias, status) {
  await check(this, 'bridge op status', async () => { const id = this.store.accountId(acctOf(this, alias).handle); const row = this.store.get('SELECT * FROM bridge_ops WHERE account_id = ? ORDER BY id DESC LIMIT 1', id); return { passed: !!row && row.status === status, detail: row ? `status ${row.status}` : 'no bridge op' }; });
});
Then('{string} has exactly one bridge op with external tx {string}', async function (alias, tx) {
  await check(this, 'one bridge op by ext_tx', async () => { const n = this.store.count('bridge_ops', 'WHERE ext_tx = ?', tx); return { passed: n === 1, detail: `count ${n}` }; });
});
Then('the transfer is refused with code {string}', async function (code) {
  await check(this, 'transfer refused', async () => ({ passed: this.api && this.api.status >= 400 && this.api.body && this.api.body.code === code, detail: this.api ? `status ${this.api.status} code ${this.api.body && this.api.body.code}` : 'no response' }));
});
Then('{string} has no {word} balance row', async function (alias, asset) {
  await check(this, 'no balance row', async () => { const id = this.store.accountId(acctOf(this, alias).handle); const row = id ? this.store.get('SELECT * FROM balances WHERE account_id = ? AND asset = ?', id, asset) : null; return { passed: !row, detail: row ? `amount ${row.amount}` : 'absent' }; });
});

File: python/be/db/steps/cex_steps.py
"""Steps for be-cex-db.feature and the shared "drive the service" account and
action steps that the API and FE features reuse. Mirror of
node/be/db/steps/cex.steps.js -- a plugin module.

Isolation is by fresh accounts and deltas, not reset. Amounts in the features
are decimal coins; coin() scales them to the 1e8 base units the store holds.
Steps that a feature uses under both `And`(Given) and `When` carry both
decorators, because pytest-bdd matches on the keyword group.
"""

from __future__ import annotations

import re
from pathlib import Path

import pytest
from pytest_bdd import given, parsers, then, when

from be.api.venues.minicex import ApiUnreachable, MiniCex
from be.db.store import DbUnreachable, Store, throwaway

cex = MiniCex()
UNREACHABLE = (DbUnreachable, ApiUnreachable)
SEED = Path(__file__).resolve().parents[4] / "services" / "mini-cex" / "db" / "seed.sql"


def coin(n):
    return round(float(n) * 1e8)


@pytest.fixture(autouse=True)
def cex_scenario(qa):
    qa.store = None
    qa.cex = cex
    qa.accounts = {}
    qa.api = None
    qa.tmp = None
    qa.last_bridge = None
    qa.bridge_deposit_req = None
    yield
    if qa.tmp is not None:
        qa.tmp.close()
    if qa.store is not None:
        qa.store.close()


def act(qa, fn):
    if qa.source_error:
        return None
    try:
        return fn()
    except UNREACHABLE as err:
        qa.source_error = str(err)
        return None


def check(qa, description, fn):
    if qa.source_error:
        qa.unobservable(description, "the source could not be reached -- " + qa.source_error)
        return
    try:
        passed, detail = fn()
    except UNREACHABLE as err:
        qa.unobservable(description, str(err))
        return
    qa.check(description, passed, detail)


def expect_throw(fn, needle):
    try:
        fn()
        return (False, "no error thrown")
    except Exception as err:  # noqa: BLE001
        ok = re.search(needle, str(err), re.I) is not None
        return (ok, ("rejected: " + str(err).split(chr(10))[0]) if ok else ("wrong error: " + str(err)))


def ensure_account(qa, alias):
    if alias in qa.accounts:
        return qa.accounts[alias]
    a = cex.create_account(alias)
    qa.accounts[alias] = a
    return a


def acct(qa, alias):
    return qa.accounts[alias]


# ---------------------------------------------------------------- Background

@given("the store is open and the service is reachable")
def store_open(qa):
    if qa.source_error:
        return
    try:
        qa.store = Store()
    except DbUnreachable as err:
        qa.source_error = str(err)
        return
    r = act(qa, lambda: cex.get("/assets"))
    if qa.source_error:
        return
    if not r or r["status"] != 200:
        qa.source_error = "mini-cex /assets returned " + str(r["status"] if r else "nothing")


# ---------------------------------------------------------------- throwaway schema

@given("a throwaway database with the schema applied")
def throwaway_schema(qa):
    qa.tmp = throwaway()
    qa.tmp.execute("INSERT INTO assets (symbol, name, kind, decimals, usd_micro) VALUES ('BTC','Bitcoin','crypto',8,1), ('USD','US Dollar','fiat',2,1)")


@given("a throwaway database with the schema and seed applied")
def throwaway_seed(qa):
    qa.tmp = throwaway()
    qa.tmp.executescript(SEED.read_text(encoding="utf-8"))


@then(parsers.re(r"the store has tables (?P<csv>.+)"))
def has_tables(qa, csv):
    def ev():
        have = set(qa.store.tables())
        want = [s.strip() for s in csv.split(",")]
        missing = [t for t in want if t not in have]
        return (not missing, "missing " + ",".join(missing) if missing else "all present")
    check(qa, "documented tables present", ev)


@then("inserting a balance with a negative amount fails a CHECK")
def neg_balance(qa):
    qa.tmp.execute("INSERT INTO accounts (id, handle, created_at) VALUES (1, 'x', 0)")
    ok, detail = expect_throw(lambda: qa.tmp.execute("INSERT INTO balances (account_id, asset, amount) VALUES (1, 'BTC', -1)"), r"CHECK|constraint")
    qa.check("negative balance rejected", ok, detail)


@then("inserting a ledger row with zero delta fails a CHECK")
def zero_delta(qa):
    qa.tmp.execute("INSERT INTO accounts (id, handle, created_at) VALUES (1, 'x', 0)")
    ok, detail = expect_throw(lambda: qa.tmp.execute("INSERT INTO ledger (account_id, asset, delta, reason, created_at) VALUES (1, 'BTC', 0, 'deposit', 0)"), r"CHECK|constraint")
    qa.check("zero delta rejected", ok, detail)


@then(parsers.parse('inserting a ledger row with reason "{reason}" fails a CHECK'))
def bad_reason(qa, reason):
    ok, detail = expect_throw(lambda: qa.tmp.execute(f"INSERT INTO ledger (account_id, asset, delta, reason, created_at) VALUES (1, 'BTC', 1, '{reason}', 0)"), r"CHECK|constraint")
    qa.check("bad reason rejected", ok, detail)


@then("inserting a balance for a missing account fails a FOREIGN KEY")
def missing_account_fk(qa):
    ok, detail = expect_throw(lambda: qa.tmp.execute("INSERT INTO balances (account_id, asset, amount) VALUES (999, 'BTC', 1)"), r"FOREIGN KEY|constraint")
    qa.check("missing account FK rejected", ok, detail)


@then("inserting two deposit bridge ops with the same external tx fails on the second")
def dup_bridge(qa):
    qa.tmp.execute("INSERT INTO accounts (id, handle, created_at) VALUES (1, 'x', 0)")
    ins = "INSERT INTO bridge_ops (account_id, asset, amount, direction, src_chain, dst_chain, status, ext_tx, created_at, updated_at) VALUES (1,'BTC',1,'deposit','ethereum','mini-cex','credited','tx1',0,0)"
    qa.tmp.execute(ins)
    ok, detail = expect_throw(lambda: qa.tmp.execute(ins), r"UNIQUE|constraint")
    qa.check("duplicate bridge ext_tx rejected", ok, detail)


@then("inserting a pair whose base equals its quote fails a CHECK")
def same_leg_pair(qa):
    ok, detail = expect_throw(lambda: qa.tmp.execute("INSERT INTO pairs (base, quote) VALUES ('BTC','BTC')"), r"CHECK|constraint")
    qa.check("same-leg pair rejected", ok, detail)


@then("applying the seed again changes no row counts")
def seed_idempotent(qa):
    def cnt():
        return (qa.tmp.execute("SELECT COUNT(*) FROM assets").fetchone()[0], qa.tmp.execute("SELECT COUNT(*) FROM pairs").fetchone()[0])
    before = cnt()
    qa.tmp.executescript(SEED.read_text(encoding="utf-8"))
    after = cnt()
    qa.check("seed is idempotent", before == after, f"assets {before[0]}->{after[0]}, pairs {before[1]}->{after[1]}")


# ---------------------------------------------------------------- shared account + action steps

@given(parsers.parse('a fresh account "{alias}"'))
def fresh_account(qa, alias):
    act(qa, lambda: ensure_account(qa, alias))


@given(parsers.parse('a fresh account "{alias}" funded with {amt:f} {asset}'))
def fresh_funded(qa, alias, amt, asset):
    def go():
        a = ensure_account(qa, alias)
        qa.api = cex.deposit(a["token"], asset, coin(amt))
    act(qa, go)


def _deposit(qa, alias, amt, asset, key=None):
    a = ensure_account(qa, alias)
    qa.api = cex.deposit(a["token"], asset, coin(amt), key)


@given(parsers.parse('"{alias}" deposits {amt:f} {asset}'))
@when(parsers.parse('"{alias}" deposits {amt:f} {asset}'))
def deposits(qa, alias, amt, asset):
    act(qa, lambda: _deposit(qa, alias, amt, asset))


@when(parsers.parse('"{alias}" deposits {amt:f} {asset} with key "{key}"'))
def deposits_key(qa, alias, amt, asset, key):
    act(qa, lambda: _deposit(qa, alias, amt, asset, key))


@when(parsers.parse('"{alias}" withdraws {amt:f} {asset}'))
def withdraws(qa, alias, amt, asset):
    def go():
        qa.api = cex.withdraw(acct(qa, alias)["token"], asset, coin(amt))
    act(qa, go)


@when(parsers.parse('"{alias}" transfers {amt:f} {asset} to "{to}"'))
def transfers(qa, alias, amt, asset, to):
    def go():
        to_acct = ensure_account(qa, to)
        qa.api = cex.transfer(acct(qa, alias)["token"], to_acct["handle"], asset, coin(amt))
    act(qa, go)


@when(parsers.parse('"{alias}" tries to transfer {amt:f} {asset} to "{to}"'))
def tries_transfer(qa, alias, amt, asset, to):
    def go():
        to_acct = ensure_account(qa, to)
        qa.api = cex.transfer(acct(qa, alias)["token"], to_acct["handle"], asset, coin(amt))
    act(qa, go)


@when(parsers.parse('"{alias}" swaps {amt:f} {frm} to {to}'))
def swaps(qa, alias, amt, frm, to):
    def go():
        qa.api = cex.swap(acct(qa, alias)["token"], {"from_asset": frm, "to_asset": to, "from_amount": coin(amt)})
    act(qa, go)


@when(parsers.parse('"{alias}" swaps {amt:f} {frm} to {to} demanding at least {minv:f} {q} out'))
def swaps_min(qa, alias, amt, frm, to, minv, q):
    def go():
        qa.api = cex.swap(acct(qa, alias)["token"], {"from_asset": frm, "to_asset": to, "from_amount": coin(amt), "min_to_amount": coin(minv)})
    act(qa, go)


@given(parsers.parse('"{alias}" bridge-withdraws {amt:f} {asset} to chain "{chain}"'))
@when(parsers.parse('"{alias}" bridge-withdraws {amt:f} {asset} to chain "{chain}"'))
def bridge_withdraw(qa, alias, amt, asset, chain):
    def go():
        qa.api = cex.bridge_withdraw(acct(qa, alias)["token"], {"asset": asset, "amount": coin(amt), "dst_chain": chain})
        if qa.api["status"] == 201:
            qa.last_bridge = qa.api["body"]
    act(qa, go)


@when(parsers.parse('a bridge deposit of {amt:f} {asset} for "{alias}" from chain "{chain}" with tx "{tx}" is observed'))
def bridge_deposit(qa, amt, asset, alias, chain, tx):
    def go():
        a = ensure_account(qa, alias)
        qa.bridge_deposit_req = {"handle": a["handle"], "asset": asset, "amount": coin(amt), "src_chain": chain, "ext_tx": tx}
        qa.api = cex.bridge_deposit(qa.bridge_deposit_req)
        if qa.api["body"] and qa.api["body"].get("id"):
            qa.last_bridge = qa.api["body"]
    act(qa, go)


@when("the same bridge deposit is observed again")
def bridge_deposit_again(qa):
    act(qa, lambda: setattr(qa, "api", cex.bridge_deposit(qa.bridge_deposit_req)))


@when("the bridge op is confirmed")
def bridge_confirm(qa):
    def go():
        qa.api = cex.bridge_confirm(qa.last_bridge["id"])
        qa.last_bridge = qa.api["body"]
    act(qa, go)


# ---------------------------------------------------------------- DB Thens

def _bal(qa, alias, asset):
    return qa.store.balance(qa.store.account_id(acct(qa, alias)["handle"]), asset)


def _ledger_sum(qa, alias, asset):
    return qa.store.ledger_balance(qa.store.account_id(acct(qa, alias)["handle"]), asset)


@then(parsers.parse('the balance row for "{alias}" in {asset} is {amt:f} {a2}'))
def balance_row(qa, alias, asset, amt, a2):
    check(qa, f"balance {alias}/{asset}", lambda: ((_bal(qa, alias, asset) == coin(amt)), f"got {_bal(qa, alias, asset)}, want {coin(amt)}"))


@then(parsers.parse('a ledger line credits "{alias}" {amt:f} {asset} with reason "{reason}"'))
def ledger_credit(qa, alias, amt, asset, reason):
    def ev():
        aid = qa.store.account_id(acct(qa, alias)["handle"])
        row = qa.store.get("SELECT * FROM ledger WHERE account_id=? AND asset=? AND delta=? AND reason=?", aid, asset, coin(amt), reason)
        return (row is not None, "found" if row else "no matching ledger line")
    check(qa, f"ledger credit {reason}", ev)


@then(parsers.parse('a ledger line debits "{alias}" {amt:f} {asset} with reason "{reason}"'))
def ledger_debit(qa, alias, amt, asset, reason):
    def ev():
        aid = qa.store.account_id(acct(qa, alias)["handle"])
        row = qa.store.get("SELECT * FROM ledger WHERE account_id=? AND asset=? AND delta=? AND reason=?", aid, asset, -coin(amt), reason)
        return (row is not None, "found" if row else "no matching ledger line")
    check(qa, f"ledger debit {reason}", ev)


@then(parsers.parse('the ledger sum for "{alias}" in {asset} equals the balance row'))
def ledger_eq_balance(qa, alias, asset):
    check(qa, f"ledger == balance {asset}", lambda: ((_ledger_sum(qa, alias, asset) == _bal(qa, alias, asset)), f"ledger {_ledger_sum(qa, alias, asset)}, balance {_bal(qa, alias, asset)}"))


@then(parsers.parse('the total BTC across "{a1}" and "{a2}" is unchanged'))
def total_conserved(qa, a1, a2):
    check(qa, "total conserved", lambda: (((_bal(qa, a1, "BTC") + _bal(qa, a2, "BTC")) == coin(1.0)), f"total {_bal(qa, a1, 'BTC') + _bal(qa, a2, 'BTC')}"))


@then(parsers.parse('"{alias}" has a swap row from {frm} to {to}'))
def swap_row(qa, alias, frm, to):
    def ev():
        aid = qa.store.account_id(acct(qa, alias)["handle"])
        row = qa.store.get("SELECT * FROM swaps WHERE account_id=? AND from_asset=? AND to_asset=?", aid, frm, to)
        return (row is not None, f"swap #{row['id']}" if row else "no swap row")
    check(qa, "swap row present", ev)


@then(parsers.parse('the bridge op for "{alias}" is in status "{status}"'))
def bridge_status_db(qa, alias, status):
    def ev():
        aid = qa.store.account_id(acct(qa, alias)["handle"])
        row = qa.store.get("SELECT * FROM bridge_ops WHERE account_id=? ORDER BY id DESC LIMIT 1", aid)
        return (row is not None and row["status"] == status, f"status {row['status']}" if row else "no bridge op")
    check(qa, "bridge op status", ev)


@then(parsers.parse('"{alias}" has exactly one bridge op with external tx "{tx}"'))
def one_bridge(qa, alias, tx):
    check(qa, "one bridge op by ext_tx", lambda: ((qa.store.count("bridge_ops", "WHERE ext_tx = ?", tx) == 1), f"count {qa.store.count('bridge_ops', 'WHERE ext_tx = ?', tx)}"))


@then(parsers.parse('the transfer is refused with code "{code}"'))
def transfer_refused(qa, code):
    check(qa, "transfer refused", lambda: ((qa.api and qa.api["status"] >= 400 and qa.api["body"] and qa.api["body"].get("code") == code), f"status {qa.api['status']} code {qa.api['body'].get('code') if qa.api['body'] else None}" if qa.api else "no response"))


@then(parsers.parse('"{alias}" has no {asset} balance row'))
def no_balance_row(qa, alias, asset):
    def ev():
        aid = qa.store.account_id(acct(qa, alias)["handle"])
        row = qa.store.get("SELECT * FROM balances WHERE account_id=? AND asset=?", aid, asset) if aid else None
        return (row is None, f"amount {row['amount']}" if row else "absent")
    check(qa, "no balance row", ev)
